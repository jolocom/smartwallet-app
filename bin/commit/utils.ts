import childProcess from 'child_process'

export const promisify =
  <T>(fn: (args: T) => void) =>
  (args: T): Promise<void> =>
    new Promise((res, rej) => {
      try {
        fn(args)
        res()
      } catch (e) {
        rej(e)
      }
    })

export const listStagedFiles = () =>
  childProcess.execSync(`git diff --cached --name-only --diff-filter=ACMR`)

export const stageModifiedFiles = (stagedFiles: string) =>
  childProcess.execSync(`echo "${stagedFiles}" | xargs git add`)

export const abortScript = (msg: string) => {
  console.log('')
  console.log('\x1b[0;33m%s\x1b[0m', msg)
  console.log('\x1b[0;31m%s\x1b[0m', 'Aborting')
  process.exit(1)
}

export const logStep = (msg: string) => {
  console.log('\x1b[0;34m%s\x1b[0m', `${msg}`)
}

export const spawnProcess = (
  onClose: (
    code: number | null,
    dataOutput: string,
    errorOutput: string,
  ) => void,
  cmd: string,
  args?: readonly string[],
  options?: childProcess.SpawnOptionsWithoutStdio,
) => {
  const spawnedProcess = childProcess.spawn(cmd, args, options)

  let dataOutput = ''
  let errorOutput = ''

  spawnedProcess.stdout.on('data', (data) => {
    dataOutput += data.toString() + '\n'
  })
  spawnedProcess.stderr.on('data', (error: Buffer) => {
    errorOutput += error.toString() + '\n'
  })

  spawnedProcess.stderr.pipe(process.stderr)
  spawnedProcess.stdout.pipe(process.stdout)

  spawnedProcess.on('close', (code) => onClose(code, dataOutput, errorOutput))

  return spawnedProcess
}
