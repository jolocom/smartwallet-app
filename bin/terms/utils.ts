import fetch from 'node-fetch';
import fs from 'fs';
import shell from 'shelljs'

const CURRENT_PATH = process.cwd();
const SECRETS_REPO = 'dev@hetz1.jolocom.io:poeditor-key'
const BASE_URL = 'https://api.poeditor.com/v2';
const API_KEY_LOCATION = `${CURRENT_PATH}/bin/poeditor-key/POEDITOR_TOKEN.txt`
const PROJECT_ID_LOCATION = `${CURRENT_PATH}/bin/poeditor-key/POEDITOR_SW2_PROJECT_ID.txt`

export enum Languages {
  en = 'en',
  de = 'de',
}

interface IResult  {
  [key: string]: {
    [key: string]: number
  }
}

export interface IResponse<T = IResult> {
  response: {
    status: string,
    code: string,
    message: string,
  },
  result: T
}

export const cloneSecrets = () => {
  return new Promise((res, rej) => {
    const keyExists = fs.existsSync(API_KEY_LOCATION);
    const idExists = fs.existsSync(PROJECT_ID_LOCATION);
    if(!keyExists || !idExists) {
      shell.cd(`${CURRENT_PATH}/bin`)
      shell.rm('-rf', `poeditor-key`)
      shell.exec(
        `git clone ${SECRETS_REPO} -b master`,
        {},
        (code, stdout, stderr) => {
          if (code != 0) return rej(new Error(stderr))
          return res(stdout)
        },
      )
    } else {
      res('Secrets are already present');
    }
  })
}

const readTextFile = (location: string) => {
  return new Promise<string>((res, rej) => {
    fs.readFile(location, 'utf8', (err, data) => {
      if (err) rej(err)
      else res(data.toString().trim())
    })
  })
}

const getPoeditorSecrets = async () => {
    const apiKey = await readTextFile(API_KEY_LOCATION)
    const projectId = await readTextFile(PROJECT_ID_LOCATION)
    return { apiKey, projectId }
}


export const sendPostRequest = async <T = IResponse, >(path: string, params: Record<string, string>) => {
  try {
    const { apiKey, projectId } = await getPoeditorSecrets();
    const requestParams = new URLSearchParams({
      api_token: apiKey,
      id: projectId,
      ...params
    })
    const res = await fetch(`${BASE_URL}${path}`, {
      method: 'POST',
      body: requestParams.toString(),
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      }
    })
    return await res.json() as T;
  } catch(err) {
    console.error(err);
  }
}