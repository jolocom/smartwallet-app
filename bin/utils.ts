import fs from 'fs'
import { stringify as stringifyEnv, parse as parseEnv, Data } from 'envfile'
import shell from 'shelljs'

export const cloneRepo = async (remote: string, repoName: string) => {
  return new Promise<string>((res, rej) => {
    const dir = shell.tempdir()
    console.log(`${dir}/${repoName}`)
    shell.rm('-rf', `${dir}/${repoName}`)
    shell.exec(
      `git clone ${remote}:${repoName} -b master ${dir}/${repoName}`,
      {},
      (code, stdout, stderr) => {
        if (code != 0) return rej(new Error(stderr))
        return res(`${dir}/${repoName}`)
      },
    )
  })
}

export const catFile = async (path: string) => {
  return new Promise((res, rej) => {
    const dsn = shell.cat(path)
    if (dsn.stderr) rej(dsn.stderr)
    res(dsn.stdout)
  })
}

export const writeEnv = async (path: string, data: object) => {
  const env = stringifyEnv(data)

  return new Promise((res, rej) => {
    fs.writeFile(path, env, (err) => {
      if (err) rej(err)
      else res(undefined)
    })
  })
}

export const readEnv = async (path: string) => {
  return new Promise<Data>((res, rej) => {
    fs.readFile(path, 'utf8', (err, data) => {
      if (err) {
        console.log(
          '\x1b[4;31m%s\x1b[0m',
          'Error! Could not find the environmental config (.env).',
        )
      } else res(parseEnv(data))
    })
  })
}
