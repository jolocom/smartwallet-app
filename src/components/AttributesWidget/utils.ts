import { AttrsState } from '~/modules/attributes/types'
import { SelectableAttrI } from './index'

export const getMappedAttrubited = (attrs: AttrsState<string>) =>
  Object.keys(attrs).reduce((acc, val) => {
    acc[val] = attrs[val].map((attr) => ({
      val: attr,
      isSelected: false,
    }))
    return acc
  }, {} as AttrsState<SelectableAttrI>)
