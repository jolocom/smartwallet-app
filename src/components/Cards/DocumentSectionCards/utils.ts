export const splitFields = <T>(fields: Array<T>) => {
  const halfway = Math.ceil(fields.length / 2)
  const last = fields.splice(halfway, fields.length)
  const first = fields.splice(0, halfway)

  return [first, last]
}
