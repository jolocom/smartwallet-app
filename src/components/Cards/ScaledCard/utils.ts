import {
  LayoutChangeEvent,
  StyleSheet,
  TextStyle,
  ViewStyle,
} from 'react-native'
import { TSupportedComponentProps } from './types'

/**
 * @internal
 */
export const scaleStyleObject = <T extends TSupportedComponentProps['style']>(
  style: T,
  scaleBy: number,
) => {
  style = StyleSheet.flatten(style) as T

  return Object.entries(style ?? {}).reduce<ViewStyle | TextStyle>(
    (acc, [key, value]) => ({
      ...acc,
      [key]: Number.isInteger(value) ? value * scaleBy : value,
    }),
    {},
  )
}

/**
 * @internal
 */
export const handleContainerLayout = (e: LayoutChangeEvent) => {
  const { width, height } = e.nativeEvent.layout
  return {
    width,
    height,
  }
}
