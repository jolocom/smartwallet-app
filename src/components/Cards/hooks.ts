import { DisplayVal } from '@jolocom/sdk/js/credentials'
import React, { useMemo, useRef, useState } from 'react'
import { TextLayoutEvent } from '~/types/props'

/**
 * logic to define if credential text should be scaled
 */
export const useCredentialNameScale = () => {
  const [isCredentialNameScaled, setIsCredentialNameScaled] = useState(false)
  const handleCredentialNameTextLayout = (e: TextLayoutEvent) => {
    if (!isCredentialNameScaled) {
      setIsCredentialNameScaled(e.nativeEvent.lines.length > 2)
    }
  }
  return {
    isCredentialNameScaled,
    handleCredentialNameTextLayout,
  }
}

/**
 * a hook to decide about nr of displayed fields for document
 * section cards
 */
export const useTrimFields = <FP>(
  fields: FP[],
  photo?: string,
  highlight?: string,
) => {
  const [displayedFields, setDisplayedFields] = useState(fields.slice(0, 3))
  const lines = useRef(0)
  const handleOptionalFieldTextLayout = () => {
    let calculatedTimes = 0
    return (e: TextLayoutEvent) => {
      calculatedTimes++
      // disable lines manipulation if the number of times this function was invoked
      // exceeds length of optional fields twice (because we calculate field name and
      // field value )
      if (calculatedTimes < fields.length * 2 + 1) {
        const numberOfLines = e.nativeEvent.lines.length
        lines.current += numberOfLines
        if (calculatedTimes === fields.length * 2) {
          /* check wether to show last optional field */
          if (lines.current > 7 && (highlight || photo)) {
            setDisplayedFields((prevState) =>
              prevState.slice(0, Math.floor(lines.current / fields.length)),
            )
          } else if (lines.current > 9 && !highlight) {
            setDisplayedFields((prevState) => prevState.slice(0, 3))
          }
        }
      }
    }
  }
  const onTextLayoutChange = handleOptionalFieldTextLayout()
  return { displayedFields, onTextLayoutChange }
}

/**
 * a hook to calculate nr of displayed fields for interaction cards
 */
export const useCalculateFieldLines = (maxLinesPerField = 2) => {
  const [fieldLines, setFieldLines] = useState<Record<number, number>>({})
  const handleFieldValueLayout = (e: TextLayoutEvent, idx: number) => {
    const lines = e.nativeEvent.lines.length
    setFieldLines((prevState) => {
      const value = prevState[idx] ?? lines
      return {
        ...prevState,
        [idx]: value > maxLinesPerField ? maxLinesPerField : value,
      }
    })
  }
  return {
    fieldLines,
    handleFieldValueLayout,
  }
}

/**
 *
 * @param fields
 * @param maxNrFields
 * @param maxNrFieldLines
 * @returns
 */

/**
 * NOTE: this hook is currently used in
 * DocumentSectionDocumentCard and InteractionShareOtherCard components.
 * @param fields - claims of a credential
 * @param maxNrField - max number of fields a card can display
 * @param maxNrFieldLines - max number of lines each field can display
 * @returns displayFields - pruned field length based on maxNrField argument
 * @returns handleFieldValueLayout - fn executed in onTextLayout event of field value text
 * to calculate how many lines each field value has
 * @returns handleFieldValuesVisibility - used to decide how many fields
 * and how many field lines in field value are displayed. Should it thould beused in FieldCalculator component
 */
export const usePruneFields = (
  fields: Array<Required<DisplayVal>>,
  maxNrFields: number,
  maxNrFieldLines: number,
) => {
  const displayedFields = useMemo(
    () => fields.splice(0, maxNrFields),
    [fields.length],
  )
  const { fieldLines, handleFieldValueLayout } = useCalculateFieldLines()

  const sumFieldLines = useMemo(() => {
    if (Object.keys(fieldLines).length === displayedFields.length) {
      return Object.keys(fieldLines).reduce(
        (acc, key) => acc + fieldLines[parseInt(key)],
        0,
      )
    }
    return 0
  }, [JSON.stringify(fieldLines)])

  /**
   * NOTE: We can not display more than maxNrFieldLines lines of all field value lines
   */
  /**
   *
   * @param child a field node (a fragement with 4 children) of the collection of displayfields node
   * children:
   * 0 idx - top padding of a field
   * 1 idx - field label
   * 2 idx - padding between field label and field value
   * 3 idx - field value
   * @param idx an index of a field node in collection of displayedFields
   * @returns
   * - a field node if sum of displayed filed lines doesnt exceeed maxNrFieldLines
   * - null if no space is left because all available field lines were occupied
   * - modiedied field node, where field value display number of remaining available fields (nr = 1)
   */
  const handleFieldValuesVisibility = (child: React.ReactNode, idx: number) => {
    /**
     * Once nr of lines of all displayed fields was calculated
     */
    if (
      Object.keys(fieldLines).length === displayedFields.length &&
      sumFieldLines !== 0
    ) {
      /**
       * if sum of all field lines doesn't exceed max
       * amount of lines that can be displayed
       */
      if (sumFieldLines <= maxNrFieldLines) {
        return child
      } else {
        /**
         * safely display lines of first and second fields,
         * because max number of lines for a field value is 2,
         * and even if both of these fields (1st, 2nd)
         * have max amount of field lines display it will still display 4 lines
         */
        if (idx === 0 || idx === 1) {
          return child
        } else {
          const remainingNrLines =
            maxNrFieldLines - fieldLines[0] - fieldLines[1]
          /**
           * If no lines are left to display do not display the whole field
           */
          if (remainingNrLines <= 0) {
            return null
          } else {
            /**
             * otherwise,
             * change numberOfLines display for field value
             */
            return React.Children.map(child.props.children, (c, idx) => {
              /**
               * NOTE: performing operation on the last child as this is a components
               * responsible for rendering field.value,
               * NOTE: !!!! if texts will get reorganized this will have to be updated
               */
              /**
               * NOTE: as of now children of field caculator fragment are:
               * 0 idx - top padding of a field
               * 1 idx - field label
               * 2 idx - padding between field label and field value
               * 3 idx - field value
               */
              if (idx === child.props.children.length - 1) {
                return React.cloneElement(c, {
                  numberOfLines:
                    maxNrFieldLines - fieldLines[0] - fieldLines[1],
                })
              }
              // the rest of children
              return c
            })
          }
        }
      }
    }
    return child
  }

  return {
    displayedFields,
    handleFieldValueLayout,
    handleFieldValuesVisibility,
  }
}
