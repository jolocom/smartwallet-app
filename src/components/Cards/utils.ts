export const getTrimmedHighlight = (highlight?: string, nrChar = 14) => {
  if (highlight) {
    return highlight?.length > nrChar
      ? highlight?.slice(0, nrChar) + '...'
      : highlight
  } else {
    return undefined
  }
}
