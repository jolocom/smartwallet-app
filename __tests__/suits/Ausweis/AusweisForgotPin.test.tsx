import React from 'react'
import {
  AusweisForgotPin,
  AUSWEIS_SUPPORT_EMAIL,
  AUSWEIS_SUPPORT_PHONE,
} from '~/screens/LoggedIn/eID/components/AusweisForgotPin'
import { useGoBack } from '~/hooks/navigation'
import { renderWithSafeArea } from '../../utils/renderWithSafeArea'
import { fireEvent } from '@testing-library/react-native'

const mockOpenUrl = jest.fn()
jest.mock('../../../src/hooks/navigation')
jest.mock('react-native/Libraries/Linking/Linking', () => ({
  openURL: mockOpenUrl,
}))

describe('Ausweis forgot pin', () => {
  beforeAll(() => {
    ;(useGoBack as jest.Mock).mockReturnValue(jest.fn)
  })
  afterEach(() => {
    mockOpenUrl.mockClear()
  })
  afterAll(() => {
    jest.resetAllMocks()
  })
  test('is displayed according to the designs', () => {
    const { toJSON } = renderWithSafeArea(<AusweisForgotPin />)
    expect(toJSON()).toMatchSnapshot()
  })
  test('user is able to navigate to email app', () => {
    const { getByText } = renderWithSafeArea(<AusweisForgotPin />)
    const contactLink = getByText(AUSWEIS_SUPPORT_EMAIL)
    fireEvent.press(contactLink)
    expect(mockOpenUrl).toHaveBeenCalledTimes(1)
    expect(mockOpenUrl).toHaveBeenCalledWith(`mailto:${AUSWEIS_SUPPORT_EMAIL}`)
  })
  test('user is able to navigate to telephone app', () => {
    const { getByText } = renderWithSafeArea(<AusweisForgotPin />)
    const phoneLink = getByText(new RegExp(AUSWEIS_SUPPORT_PHONE))
    fireEvent.press(phoneLink)
    expect(mockOpenUrl).toHaveBeenCalledTimes(1)
    expect(mockOpenUrl).toHaveBeenCalledWith(`tel:${AUSWEIS_SUPPORT_PHONE}`)
  })
})
