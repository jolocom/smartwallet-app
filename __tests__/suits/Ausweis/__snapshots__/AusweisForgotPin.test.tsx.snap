// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ausweis forgot pin is displayed according to the designs 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        Object {
          "flex": 1,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "justifyContent": "center",
          },
          false,
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#101010",
            "height": 0,
            "width": "100%",
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 50,
              "paddingHorizontal": 12,
              "width": "100%",
            },
            Object {
              "justifyContent": "space-between",
            },
            Array [
              Object {
                "backgroundColor": "#101010",
              },
              Object {},
            ],
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "width": 40,
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={17}
            bbWidth={25}
            focusable={false}
            height={17}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 17,
                  "width": 25,
                },
              ]
            }
            vbHeight={17}
            vbWidth={25}
            width={25}
          >
            <RNSVGGroup>
              <RNSVGPath
                d="M24.48 7.438H3.34l5.362-5.469a.536.536 0 00.152-.375V.53A.53.53 0 008.334 0a.516.516 0 00-.369.156L.153 8.124a.539.539 0 000 .752l7.812 7.968c.15.152.372.198.568.116a.53.53 0 00.321-.491v-1.063a.536.536 0 00-.152-.375L3.34 9.562h21.138c.288 0 .521-.237.521-.53V7.968a.526.526 0 00-.52-.531z"
                fill={-855310}
                fillRule={1}
                propList={
                  Array [
                    "fill",
                    "fillRule",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "paddingHorizontal": 10,
            }
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "paddingHorizontal": 16,
              "width": "100%",
            },
            Object {
              "backgroundColor": "#101010",
              "marginBottom": -0,
              "paddingBottom": 0,
            },
            Object {},
            Object {
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "paddingTop": undefined,
                "textAlign": "center",
              },
              Object {
                "color": "rgba(255,255,255,0.85)",
                "fontFamily": "TTCommons-Regular",
                "fontSize": 28,
                "letterSpacing": 0,
                "lineHeight": 32,
              },
              Object {
                "paddingHorizontal": 28,
              },
            ]
          }
        >
          AusweisLostPin.title
        </Text>
        <Text
          style={
            Array [
              Object {
                "paddingTop": undefined,
                "textAlign": "center",
              },
              Object {
                "color": "rgb(138, 138, 145)",
                "fontFamily": "TTCommons-Regular",
                "fontSize": 20,
                "letterSpacing": 0.14,
                "lineHeight": 22,
              },
              Object {
                "marginTop": 24,
              },
            ]
          }
        >
          AusweisLostPin.description
        </Text>
        <Text
          style={
            Array [
              Object {
                "paddingTop": undefined,
                "textAlign": "center",
              },
              Object {
                "color": "rgb(138, 138, 145)",
                "fontFamily": "TTCommons-Regular",
                "fontSize": 20,
                "letterSpacing": 0.14,
                "lineHeight": 22,
              },
              Object {
                "marginTop": 32,
              },
            ]
          }
        >
          AusweisLostPin.contact
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "paddingTop": undefined,
                  "textAlign": "center",
                },
                Object {
                  "color": "#5251c1",
                  "fontFamily": "TTCommons-Regular",
                  "fontSize": 20,
                  "letterSpacing": 0.14,
                  "lineHeight": 22,
                },
                Object {
                  "marginTop": 32,
                },
              ]
            }
          >
            support@ausweissapp.de
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "paddingTop": undefined,
                  "textAlign": "center",
                },
                Object {
                  "color": "#5251c1",
                  "fontFamily": "TTCommons-Regular",
                  "fontSize": 20,
                  "letterSpacing": 0.14,
                  "lineHeight": 22,
                },
                Object {
                  "marginTop": 12,
                },
              ]
            }
          >
            0421 204 95 995 (AusweisLostPin.phoneRate)
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
